Documentation/replacing-component.md

```diff
@@ -0,0 +1,266 @@
+
+```swift
+protocol EndpointNode {
+    associatedtype Response: ResponseEncodable
+    
+    func handle() -> Response
+}
+
+
+protocol EndpointProvidingNode {
+    associatedtype Content: EndpointProvidingNode
+    
+    @EndpointProvidingNodeBuilder
+    var content: Content { get }
+}
+```
+
+- `EndpointNode`s are nodes which expose some functionality to the client.  
+  All leaves in the `WebService` must be `EndpointNode`s.
+- `EndpointProvidingNode`s are nodes which sit above the leaves.  
+  They cannot provide any functionality of their own, but they must eventually lead to one or more `EndpointNode`s.
+
+
+**Note** The names for these protocols aren't fixed, I just went with the most descriptive thing i could think of.
+
+
+
+
+### How does it affect the current implementation?
+
+Switching to this new structure required surprisingly few changes.
+In most cases it was sufficient to replace `Component` with `EndpointNode`, since the semantic model builder already operates only on endpoint nodes, and never on non-endpoint nodes. All tests still pass.
+
+- `Text` is now an `EndpointNode`
+- `EmptyComponent` is removed, since user-exposed empty components were invalid DSL components to begin with.  
+  (It still is possible to define a noop component, by typealiasing `Content` or `Response` to `Never`.)
+- Some tests needed to be updated (since a component's nested handlers can't be accessed via `component.content as? SomeHandlerType`)
+- `AnyComponent` is replaced by `AnyEndpointNode` and `AnyEndpointProvidingNode`
+- Neither `content` nor `handle()` have a default implementation anymore.
+  You have to provide an implementation for your components to compile.
+  (The exception here is if you manually add a `typealias {Content|Response} = Never`, in which case we do still provide a default implementation. The idea is that, if you want a component to be unusable, and crash if ever accessed, you have to explicitly specify this.)
+
+
+
+### Benefits from switching to a more fine-grained protocol structure:
+
+**Type Safety**
+
+With the current implementation, we'd need to rely heavily on runtime checks to validate the DSL structure and rejct invalid web services.
+
+By using protocols to define some of these component concepts, we can make use of the type system to enforce these requirements.
+
+This directly benefits both the user (ie the person implementing a web service), and us:
+
+- The user will, in some cases, get compile-time errors when trying to compile incorrectly implemnented components, or ill-formed DSL constructs
+- We can, when dealing with DSL nodes, always know for a fact their specific type and functionalities and won't have to do any dynamic or runtime-metadata based checks to determine that information
+
+
+
+
+### Issues with the current implementation in [experimental/dsl-node-protocols][branch_url]
```

Comment:
Would this be solved if `Handler`/`EndpointNode` is also conforming to `Component`/`EndpointProvidingNode` as described here https://github.com/Apodini/Apodini/pull/26#discussion_r538396896? How would this work with protocols that further refine `Handler`/`EndpointNode` e.g. to enable communication between `Handler`s?

Reply:
Probably. The thing is, if we have `Handler` conform to `Component`, we end up in the same situation we currently have: all handlers might also be components, but we don't really know for sure (at least not statically).

The way i see it, the only reason for having `Handler` conform to `Component` would be being able to register endpoints on intermediate paths:
```
- GET /v1/foo
- GET /v1/foo/bar
```

This is already possible without having `Handler` conform to `Component`, by simply having a `Handler` and a `Component` on the same level w/in the DSL (in fact we're already doing this in the TestWebService).


Another issue is that, if handlers were allowed to provide further components, the reasonable expextation would be that the `content` returned by the handler introduces a new path component (ie, `content` returns a `Group`).
Otherwise, if a handler's content property returned another handler, that handler might conflict with the handler itself.